src/operator/tensor/indexing_op.h

```diff
@@ -323,58 +323,97 @@ inline bool TakeOpShape(const nnvm::NodeAttrs& attrs,
 }
 
 inline bool TakeOpType(const nnvm::NodeAttrs& attrs,
-                       std::vector<int> *in_type,
-                       std::vector<int> *out_type) {
-  // using single dtype ("float32") for safety reason
-  CHECK_GE(in_type->size(), 2U);
-  int dtype = (*in_type)[1];
-  CHECK_NE(dtype, -1) << "idx must have specified type";
-  for (index_t i = 0; i < in_type->size(); ++i) {
-    if ((*in_type)[i] == -1) {
-      (*in_type)[i] = dtype;
-    } else {
-      CHECK_EQ((*in_type)[i], dtype) << "This layer requires uniform type. "
-                                     << "Expected " << dtype << " v.s. given "
-                                     << (*in_type)[i];
-    }
+                       std::vector<int> *in_attrs,
+                       std::vector<int> *out_attrs) {
+  CHECK_EQ(in_attrs->size(), 2U);
+  CHECK_EQ(out_attrs->size(), 1U);
+  if (-1 == (*in_attrs)[1]) {
+    (*in_attrs)[1] = mshadow::kInt32;
```

Comment:
I'm thinking if it would be better to leave this as -1 rather than assign int32. Because we support backward inference. It could be filled by input op in later iterations.